datafusion/physical-plan/src/sorts/merge.rs

```diff
@@ -533,3 +534,134 @@ impl<C: CursorValues + Unpin> RecordBatchStream for SortPreservingMergeStream<C>
         Arc::clone(self.in_progress.schema())
     }
 }
+
+/// Fixed size queue implemented as a circular buffer
+/// The underlying `values` are immutable, so removing elements will not drop them and not change the
+/// capacity/length of the actual vector
+#[derive(Debug, Default)]
+struct FixedSizeQueue {
+    values: Vec<usize>,
+    start: usize,
+    end: usize,
+    len: usize,
+}
+
+impl From<Vec<usize>> for FixedSizeQueue {
+    fn from(values: Vec<usize>) -> Self {
+        let len = values.len();
+
+        Self {
+            values,
+            start: 0,
+            end: len.saturating_sub(1),
+            len,
+        }
+    }
+}
+
+impl FixedSizeQueue {
+    /// Get the value at the top of the queue
+    ///
+    /// # Implementation
+    /// return the value at [`Self::start`]
+    ///
+    /// ## Example
+    ///
+    /// ```plain
+    /// index: | 0  |  1  |  2  |  3  |  4  |  5  |  6  |  7  |  8  |  9  |
+    ///        | -------------------------------------------------------- |
+    /// value: | 0  |  1  |  2  |  3  |  4  |  5  |  6  |  7  |  8  |  9  |
```

Comment:
I would find these examples easier to follow if `index` and `value` didn't have the same values. maybe something like

```suggestion
    /// value: | 9  |  8  |  7  |  5  |  5  |  4  |  3  |  2  |  1  |  0  |
```